File: can_construct_word/js_impl/canConstructWord.js
const canConstruct = (target, wordBank) => {
    if (target === "") {
        return true
    }

    for (let word of wordBank) {
        if (target.indexOf(word) === 0) {
            const suffix = target.slice(word.length);
            if (canConstruct(suffix, wordBank) == true) {
                return true
            }
        }
    }
    return false
}


const memoCanConstruct = (target, wordBank, memo = {}) => {
    if (target in memo) return memo[target];
    if (target === "") {
        return true;
    }

    for (let word of wordBank) {
        if (target.indexOf(word) === 0) {
            const suffix = target.slice(word.length);
            if (memoCanConstruct(suffix, wordBank, memo) == true) {
                memo[target] = true;
                return true;
            }
        }
    }
    memo[target] = false;
    return false;
}



// console.log(canConstruct("abcdef", ["ab", "abc", "cd", "def", "abcd"]));
// console.log(canConstruct("skateboard", [
//     "bo", "rd", "ate", "t", "ska", "sk", "boar"
// ]));

// console.log(canConstruct("enterapotentpot", [
//     "a", "p", "en", "enter", "ot", "o", "t"
// ]))
// console.log(canConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef",
//     ["e", "ee", "eee", "eeee", "eeeee", "eeeeee"]))


console.log(memoCanConstruct("abcdef", ["ab", "abc", "cd", "def", "abcd"]));
console.log(memoCanConstruct("skateboard", [
    "bo", "rd", "ate", "t", "ska", "sk", "boar"
]));

console.log(memoCanConstruct("enterapotentpot", [
    "a", "p", "en", "enter", "ot", "o", "t"
]))
console.log(memoCanConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef",
    ["e", "ee", "eee", "eeee", "eeeee", "eeeeee"]))

File: tabulation/can_construct_word/python_impl/can_construct_word.py
def can_construct_word(target, word_bank):
    table = [False]*(len(target)+1)
    table[0] = True
    for i, _ in enumerate(target):
        if table[i] == True:
            for word in word_bank:
                if target[i:i+len(word)] == word:
                    table[i+len(word)] = True
    return table[len(target)]


print(can_construct_word("abcdef", ["ab", "abc", "cd", "def", "abcd"]))
print(can_construct_word("skateboard", [
      "bo", "rd", "ate", "t", "ska", "sk", "boar"]))
print(can_construct_word("enterapotentpot", [
      "a", "p", "en", "enter", "ot", "o", "t"]))
print(can_construct_word("eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef",
                         ["e", "ee", "eee", "eeee", "eeeee", "eeeeee"]))
